### ISSUE TITLE
Accuracy issue for BERT-large Squad task

### ISSUE THREAD

**Issue by zqliu19** [aws/aws-neuron-sdk#301]
Hi AWS experts,

We tried running BERT-large (question-answer task with Squad dataset) on inf1.2xlarge with weights obtained from MLperf inference repo, and found that the avergae F1 score we got is ~89.665%, which is lower than MLperf standard 90.874%.

Any ideas how to improve the accuracy to match the industry standard?
How do we check which datatype is used for each layer in BERT? And is it possible to adjust the data type to achieve higher F1 score?
Any compiler related optimizations (e.g., approximate computation) that might affect accuracy?

Thanks a lot!

**Comment by micwade-aws** [comment]
@zqliu19 - What is the framework you're using?  Also, if it's possible to share your sample code either here or with aws-neuron-support@amazon.com, that would help us understand what might be going on.  

Without knowing more, it's possible that by telling the compiler to use a higher precision cast instead of our default may improve accuracy of the resulting model.  More details on our default settings with mixed precision can be found [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/perf/performance-tuning.html#mixed-precision).  

Flags to pass to the compiler:
`--fp32-cast=matmult`
If your model's dynamic range allows: `--fp32-cast=matmult-fp16`

**Comment by aws-joshim** [comment]
@zqliu19 checking if you had a chance to try out the --fp32-cast=matmult-fp16 cast flag ?

**Comment by zqliu19** [comment]
> @zqliu19 checking if you had a chance to try out the --fp32-cast=matmult-fp16 cast flag ?

Thanks a lot for the suggestions.
Tried  --fp32-cast=matmult-fp16  and we actually got slightly worse overall F1 score compared to default option.
Will also try --fp32-cast=matmult

**Comment by zqliu19** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/allliu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/allliu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



pytorch | default | matmult-bf16 | matmult | matmult-fp16
-- | -- | -- | -- | --
F1 score | 89.665% | 89.634% | 89.694% | 89.650%



</body>

</html>
@aws-joshim Could you help check and answer questions below? Thanks a lot.

1. We saw Tensorflow F1 score (based on https://github.com/aws/aws-neuron-sdk/blob/master/src/examples/tensorflow/bert_demo/bert_no_model.py) is >0.2% lower than pytorch. What could be the reason for this?
2. We saw --fp32-cast=matmult has the best accuracy. What is the precision for this option and what is the difference between  matmult and  matmult-bf16/ matmult-fp16
3. default option actually has better accuracy (higher F1 score) compared to  matmult-bf16, is this expected?
4. the best F1 score (with --fp32-cast=matmult ) is still lower than industry standard (90.874%). Any suggestions how to interpret and improve this?
5. We noticed that int8 is still not supported in neuron SW, is that because INT8 has more severe accuracy issues that has not been fixed on inf1? 
Thanks a lot.

**Comment by zqliu19** [comment]
It is out evaluation methodology issue, accuracy for default cast option is ~ 91% after we fix the evaluation methodology.
Thanks for checking and closing this.
